SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -3531,6 +3533,133 @@ bool USpatialNetDriver::HasTimedOut(const float Interval, uint64& TimeStamp)
+		check(Actor != nullptr);
+
+		Worker_EntityId EntityId = PackageMap->GetEntityIdFromObject(Actor);
+
+		// Check for unauthorised data changes no non-auth actors and add auth actors only to the consider list
+		if (bSpatialNetworking && Actor->HasAuthority())
+		{
+			OutConsiderList.Add(ActorInfo);
+		}
+		else if (bSpatialNetworking && bEnableSpatialDataDebugger)
+		{
+			CheckUnauthorisedDataChanges(EntityId, Actor);
+		}
+	}
+}
+
+void USpatialNetDriver::CheckUnauthorisedDataChanges(Worker_EntityId_Key EntityId, AActor* Actor)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer() || !SpatialShadowActors.Contains(EntityId))
+	{
+		return;
+	}
+
+	if (SpatialShadowActors[EntityId] != nullptr)
+	{
+		SpatialShadowActors[EntityId]->CheckUnauthorisedDataChanges(EntityId, Actor);
+	}
+	else
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Trying to check SpatialShadowActor which is invalid, EntityID %i"), EntityId);
+		return;
+	}
+}
+
+void USpatialNetDriver::AddSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer())
+	{
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId));
+	if (Actor == nullptr || !IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		return;
+	}
+
+	if (SpatialShadowActors.Contains(EntityId))
+	{
+		UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Should only be adding a SpatialShadowActor once for each entity, EntityID %i"),
+			   EntityId);
+	}
+	else // if (!HasServerAuthority(EntityId))
+	{
+		USpatialShadowActor* SpatialShadowActor(NewObject<USpatialShadowActor>());
+		SpatialShadowActor->Init(EntityId, Actor);
+		SpatialShadowActors.Add(EntityId, SpatialShadowActor);
+	}
+}
+
+void USpatialNetDriver::RemoveSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer() || !SpatialShadowActors.Contains(EntityId))
+	{
+		return;
+	}
+
+	USpatialShadowActor* SpatialShadowActor = SpatialShadowActors[EntityId];
+	SpatialShadowActors.Remove(EntityId);
+}
+
+void USpatialNetDriver::UpdateSpatialShadowActor(Worker_EntityId_Key EntityId)
+{
+	const bool bEnableSpatialDataDebugger = GetDefault<USpatialGDKSettings>()->IsSpatialDataDebuggerEnabled();
+
+	if (!bEnableSpatialDataDebugger || !IsServer())
+	{
+		return;
+	}
+
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId));
+	if (Actor == nullptr || !IsValid(Actor) || Actor->IsPendingKillOrUnreachable())
+	{
+		return;
+	}
+
+	if (!SpatialShadowActors.Contains(EntityId))
+	{
+		// We can receive updates without receiving adds - in this case create the SpatialShadowActor
+		AddSpatialShadowActor(EntityId);
```

Comment:
this function finds the 'actor pointer', and AddSpatialShadowActor does the same?